ibis/backends/tests/test_join.py

```diff
@@ -1,81 +1,153 @@
+import numpy as np
 import pandas as pd
 import pytest
 from pytest import param
 
 
+def _pandas_semi_join(
+    left,
+    right,
+    on,
+    how="semi",
+    suffixes=("", "_y"),
+):
+    assert len(on) == 1, str(on)
+    inner = pd.merge(left, right, how="inner", on=on)
+    filt = left.loc[:, on[0]].isin(inner.loc[:, on[0]])
+    return left.loc[filt, :]
+
+
+def _pandas_anti_join(
+    left,
+    right,
+    on,
+    how="anti",
+    suffixes=("", "_y"),
+):
+    assert len(on) == 1, str(on)
+    inner = pd.merge(left, right, how="inner", on=on)
+    filt = left.loc[:, on[0]].isin(inner.loc[:, on[0]])
+    return left.loc[~filt, :]
+
+
+def _merge(
+    left,
+    right,
+    on,
+    *,
+    how,
+    suffixes=("", "_y"),
+):
+    joined = pd.merge(
+        left,
+        right,
+        on=on,
+        how=how,
+        suffixes=suffixes,
+        indicator=True,
+    )
+    if how == "right":
+        filt_key = {"left_only"}
+    elif how == "left":
+        filt_key = {"right_only"}
+    else:
+        filt_key = set()
+
+    joined.loc[joined["_merge"].isin(filt_key), on] = np.nan
+    return joined.drop(["_merge"], axis=1)
+
+
+IMPLS = {
+    "semi": _pandas_semi_join,
+    "anti": _pandas_anti_join,
+}
+
+
+def check_eq(left, right, how, **kwargs):
+    impl = IMPLS.get(how, _merge)
+    return impl(left, right, how=how, **kwargs)
+
+
 @pytest.mark.parametrize(
     "how",
     [
+        param("inner", marks=pytest.mark.notimpl(["datafusion"])),
+        param("left", marks=pytest.mark.notimpl(["datafusion"])),
         param(
-            "inner",
-            marks=pytest.mark.notimpl(["datafusion", "impala"]),
-        ),
-        param(
-            "left",
-            marks=pytest.mark.notimpl(["datafusion", "impala"]),
-        ),
-        param(
-            "right",
-            marks=pytest.mark.notimpl(
-                [
-                    "clickhouse",
-                    "datafusion",
-                    "mysql",
-                    "postgres",
-                    "sqlite",
-                    "impala",
-                ]
-            ),
+            "right", marks=pytest.mark.notimpl(["clickhouse", "datafusion"])
         ),
         param(
             "outer",
+            # TODO: mysql and sqlite will likely never support full outer join
+            # syntax, but we might be able to work around that using
+            # LEFT JOIN UNION RIGHT JOIN
             marks=pytest.mark.notimpl(
-                [
-                    "clickhouse",
-                    "datafusion",
-                    "mysql",
-                    "postgres",
-                    "sqlite",
-                    "impala",
-                ]
+                ["clickhouse", "datafusion", "mysql", "sqlite"]
```

Comment:
Haven't you [added](https://github.com/ibis-project/ibis/pull/3501/files#diff-ecfec13cc30bc29a47126b496dc07567347a41c67922abb7ad0cc7e996eb7f72R62) support for clickhouse?